services/continuousquery/service.go

```diff
@@ -0,0 +1,309 @@
+	q, ok := stmt.(*influxql.CreateContinuousQueryStatement)
+	if !ok || q.Source.Target == nil || q.Source.Target.Measurement == nil {
+		// q.Source.Target (destination) for the result of a SELECT INTO query.
+		return nil, errors.New("invalid continuous query")
+	}
+	
+	// Create a new ContinuousQuery object.
+	cq := &ContinuousQuery{
+		Database: database,
+		Info:     cqi,
+		q:        q.Source,
+	}
+	
+	return cq, nil
+}
+
+// shouldRunContinuousQuery returns true if the CQ should run.
+func (cq *ContinuousQuery) shouldRunContinuousQuery(now time.Time, interval time.Duration) (bool, time.Time, error) {
+	// check whether the query is an aggregate query.
+	if cq.q.IsRawQuery { // cq.q.IsRawQuery is true when is not an aggregate query.
+		return false, cq.LastRun, errors.New("continuous queries must be aggregate queries")
+	}
+	
+	// Determine if we should run the continuous query based on the last time it ran.
+	if cq.HasRun { // the query has run before.
+		// Retrieve the zone offset for the previous window.
+		nextRun := cq.LastRun.Add(interval)
+		return true, nextRun, nil
+	} else { // the query never ran before.
+		// Retrieve the location from the CQ.
+		loc := cq.q.Location
+		if loc == nil {
+			loc = time.UTC
+		}
+		// If the query has never run, execute it using the current time.
+		return true, now.In(loc), nil
+	}
+}
+
+// hasContinuousQuery returns true if any CQs exist.
+func (s *Service) hasContinuousQuery() bool {
+	// Get list of all databases.
+	dbs := s.MetaClient.Databases()
+	// Loop through all databases executing CQs.
+	for _, db := range dbs {
+		if len(db.ContinuousQueries) > 0 {
+			return true
+		}
+	}
+	return false
+}
+
+// runContinuousQueryAndWriteResult will run the query against the cluster and write the results back in
+func (s *Service) runContinuousQueryAndWriteResult(cq *ContinuousQuery) *query2.Result {
+	// Wrap the CQ's inner SELECT statement in a Query for the Executor.
+	q := &influxql.Query{
+		Statements: influxql.Statements([]influxql.Statement{cq.q}),
+	}
+	
+	closing := make(chan struct{})
+	defer close(closing)
+	
+	var qDuration *statistics.SQLSlowQueryStatistics
+	if !isInternalDatabase(cq.Database) {
+		qDuration = statistics.NewSqlSlowQueryStatistics()
+		qDuration.SetDatabase(cq.Database)
+		defer func() {
+			d := time.Since(time.Now())
+			if d.Nanoseconds() > time.Second.Nanoseconds()*10 {
+				qDuration.AddDuration("TotalDuration", d.Nanoseconds())
+				statistics.AppendSqlQueryDuration(qDuration)
+			}
+			s.Logger.Info("sql query duration", zap.Duration("duration", d))
+		}()
+	}
+	
+	traceId := tsi.GenerateUUID()
+	opts := query.ExecutionOptions{
+		Database:       cq.Database,
+		Chunked:        true,
+		Quiet:          true,
+		InnerChunkSize: DefaultInnerChunkSize,
+		Traceid:        traceId,
+	}
+	// Execute the SELECT INTO statement.
+	// TODO: something wrong with the query executor.
+	ch := s.QueryExecutor.ExecuteQuery(q, opts, closing, qDuration)
```

Comment:
`s.QueryExecutor` is nil, it has not been initialized.

Reply:
Thx bro.